src/render.rs

```diff
@@ -475,37 +534,161 @@ impl Renderer {
             };
 
             //TODO: batch per PSO
-            let (pso, color, param0, map) = match *material {
-                Material::LineBasic { color } => (&self.pso_line_basic, color, 0.0, None),
-                Material::MeshBasic { color, ref map, wireframe: false } => (&self.pso_mesh_basic_fill, color, 0.0, map.as_ref()),
-                Material::MeshBasic { color, map: ref _map, wireframe: true } => (&self.pso_mesh_basic_wireframe, color, 0.0, None),
-                Material::MeshLambert { color, flat } => (&self.pso_mesh_gouraud, color, if flat {0.0} else {1.0}, None),
-                Material::MeshPhong { color, glossiness } => (&self.pso_mesh_phong, color, glossiness, None),
-                Material::Sprite { ref map } => (&self.pso_sprite, !0, 0.0, Some(map)),
-            };
-            let uv_range = match map {
-                Some(ref map) => map.get_uv_range(),
-                None => [0.0; 4],
-            };
-            self.encoder.update_constant_buffer(&gpu_data.constants, &Locals {
-                mx_world: Matrix4::from(node.world_transform).into(),
-                color: decode_color(color),
-                mat_params: [param0, 0.0, 0.0, 0.0],
-                uv_range,
-            });
-            //TODO: avoid excessive cloning
-            let data = pipe::Data {
-                vbuf: gpu_data.vertices.clone(),
-                cb_locals: gpu_data.constants.clone(),
-                cb_lights: self.light_buf.clone(),
-                cb_globals: self.const_buf.clone(),
-                tex_map: map.unwrap_or(&self.map_default).to_param(),
-                shadow_map0: (shadow0.clone(), shadow_sampler.clone()),
-                shadow_map1: (shadow1.clone(), shadow_sampler.clone()),
-                out_color: self.out_color.clone(),
-                out_depth: self.out_depth.clone(),
+            match *material {
```

Comment:
This match expression is bloated. It may be worth splitting up the cases in the future.